pkg/builder/builder.go

```diff
@@ -75,19 +90,79 @@ func (b Builder) Build(
 	if err != nil {
 		return nil, err
 	}
+	buildOpts.Def = def
 
-	handler, err := b.Registry.FindHandler(def.Kind)
+	state, img, err := b.build(ctx, solver, buildOpts)
 	if err != nil {
 		return nil, err
 	}
-	handler.WithSolver(solver)
 
-	buildOpts.Def = def
+	return solveStateWithImage(ctx, c, state, img)
+}
+
+func (b Builder) build(
+	ctx context.Context,
+	solver statesolver.StateSolver,
+	buildOpts builddef.BuildOpts,
+) (llb.State, *image.Image, error) {
+	webserverStage := isWebserverStage(buildOpts.Stage)
+	if webserverStage {
+		buildOpts.Stage = strings.TrimPrefix(buildOpts.Stage, "webserver-")
+	}
+
+	handler, err := b.findHandler(buildOpts.Def.Kind, solver, webserverStage)
+	if err != nil {
+		return llb.State{}, nil, err
+	}
+
 	state, img, err := handler.Build(ctx, buildOpts)
 	if err != nil {
-		return nil, err
+		return state, img, err
+	}
+
+	if webserverStage {
+		buildOpts.Def = newBuildDefForWebserver(buildOpts.Def)
+		buildOpts.Source = &state
+		buildOpts.Stage = "webserver"
```

Comment:
Why to set again this value whereas it was already set on line 110 ?
Is line 110 still relevant ?

Reply:
The `webserver-` prefix is removed from stage name at line 110 (`webserver-prod` -> `prod`) :wink: